pyomo/contrib/incidence_analysis/interface.py

```diff
@@ -101,10 +111,36 @@ def get_bipartite_incidence_graph(variables, constraints, include_fixed=True):
     graph.add_nodes_from(range(M), bipartite=0)
     graph.add_nodes_from(range(M, M + N), bipartite=1)
     var_node_map = ComponentMap((v, M + i) for i, v in enumerate(variables))
-    for i, con in enumerate(constraints):
-        for var in identify_variables(con.expr, include_fixed=include_fixed):
-            if var in var_node_map:
-                graph.add_edge(i, var_node_map[var])
+    vars_in_cons = list(
+        _generate_variables_in_constraints(constraints, include_fixed=True)
+    )
+    to_reset = [
+        var for var in vars_in_cons if var.fixed and var.value is None
+    ]
+    if include_fixed:
+        # AMPLRepnVisitor will not include fixed variables, so if we need to
+        # include these variables, we temporarily unfix them to generate the
+        # AMPLRepn, then re-fix them when we are done.
+        to_unfix = [var for var in variables if var.fixed]
+        context = TemporarySubsystemManager(to_unfix=to_unfix, to_reset=to_reset)
+    else:
+        # Note that AMPLRepnVisitor evaluates the constant term in an
+        # expression, including fixed variables. If any fixed variable
+        # does not have a value, this will fail. Since the constant term
+        # is irrelevant for the purpose of an incidence graph, we will
+        # temporarily assign these uninitialized fixed variables a value of
+        # 0.0, then reset them to None after generating the incidence graph.
+        context = TemporarySubsystemManager(to_reset=to_reset)
+
+    with context:
+        for var in to_reset:
+            # We use random.uniform here to defend against arbitrary fixed
+            # variable values causing a numeric cancellation.
+            var.set_value(random.uniform(0.0, 1.0))
```

Comment:
Setting the values of `None`-valued fixed variables can cause us to generate a "wrong" incidence graph if the values we choose happen to cause a numeric cancellation. The obvious values of zero and one happen to be particularly likely to cause a cancellation via terms like `x*y` and `(1-x)*y`. This is a horrible hack but leads to a vanishingly small probability of introducing a numeric cancellation. Alternative suggestions welcome.

Reply:
We could get around using `random.uniform` here if `AMPLRepnVisitor` did not replace `NaN*x -> NaN`. In this case, the `AMPLRepn` linear portion would look like `{id(x): float('nan')`. In this case, we could temporarily set fixed variables (and mutable parameters) to NaN to guarantee no cancellation while generating the `AMPLRepn`.